app/code/Magento/Inventory/Controller/Adminhtml/Stock/StockSourceLinkProcessor.php

```diff
@@ -77,25 +78,24 @@ public function process(int $stockId, array $stockSourceLinksData)
     {
         $this->validateStockSourceData($stockSourceLinksData);
 
-        $assignedSources = $this->getAssignedSourcesForStock->execute($stockId);
+        $assignedLinks = $this->getAssignedLinks($stockId);
         $sourceCodesForSave = array_flip(array_column($stockSourceLinksData, StockSourceLink::SOURCE_CODE));
         $sourceCodesForDelete = [];
 
-        foreach ($assignedSources as $assignedSource) {
-            if (array_key_exists($assignedSource->getSourceCode(), $sourceCodesForSave)) {
-                unset($sourceCodesForSave[$assignedSource->getSourceCode()]);
+        foreach ($assignedLinks as $assignedLink) {
+            if (array_key_exists($assignedLink->getSourceCode(), $sourceCodesForSave)) {
+                unset($sourceCodesForSave[$assignedLink->getSourceCode()]);
```

Comment:
Currently, it looks non intuitive, because you get Links objects, after that you iterate over that array and extract source codes, and after that based on that source codes build Links again.

Why do we need this intermediate step at all? 

All that we need here - to work just with Link objects. That's all